src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -687,5 +690,70 @@ private class ClassWithMultipleDictionaries
 
             public Dictionary<string, object> ActualDictionary { get; set; }
         }
+
+        [Fact]
+        public static void CustomObjectConverterInExtensionProperty()
+        {
+            const string Json = "{\"hello\": \"world\"}";
+
+            var options = new JsonSerializerOptions();
+            options.Converters.Add(new JsonObjectConverter());
+
+            ClassWithExtensionPropertyAsObject obj = JsonSerializer.Deserialize<ClassWithExtensionPropertyAsObject>(Json, options);
+            object overflowProp = obj.MyOverflow["hello"];
+            Assert.IsType<string>(overflowProp);
+            Assert.Equal("world!!!", ((string)overflowProp));
+
+            string newJson = JsonSerializer.Serialize(obj, options);
+            Assert.Equal("{\"hello\":\"world!!!\"}", newJson);
+        }
+
+        private class JsonObjectConverter : JsonConverter<object>
+        {
+            public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                return reader.GetString() + "!!!";
+            }
+
+            public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+            {
+                // Since we are converter for object, the Hashtable converter will be called instead of this.
+                throw new InvalidOperationException();
```

Comment:
Btw, we aren't calling the user-defined `JsonConverter<object>.Write` when the runtime type is actually `System.Object`. This behavior exist on 3.1 too.

```C#
public static void Main()
{
    var obj = new object();
    var options = new JsonSerializerOptions { Converters = { new JsonObjectConverter(), new JsonStringConverter() } };
    string json = JsonSerializer.Serialize(obj, options);
    Console.WriteLine(json); // {}, doesn't call JsonObjectConverter.Write, it should throw

    obj = "foo";
    json = JsonSerializer.Serialize(obj, options); // this does call JsonStringConverter.Write and throws
    Console.WriteLine(json);
}

private class JsonObjectConverter : JsonConverter<object>
{
    public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }

    public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}

private class JsonStringConverter : JsonConverter<string>
{
    public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }

    public override void Write(Utf8JsonWriter writer, string value, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }
}
```

Reply:
We can change this, but not sure who would want to override System.Object to writing anything other than "{}"...